src/support/qsort.c

```diff
@@ -0,0 +1,263 @@
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include "wt_internal.h"
+
+#define WT_QSORT_INSERT_THRESHOLD 7
+#define WT_QSORT_LARGE_THRESHOLD 40
+
+#ifdef WT_QSORT_R_DEF
+typedef int wt_cmp_t(const void *, const void *, void *);
+#else
+typedef int wt_cmp_t(const void *, const void *);
+#endif
+
+#ifdef WT_QSORT_R_DEF
+#define CMP(x, y, t) (cmp((x), (y), (t)))
+#else
+#define CMP(x, y, t) (cmp((x), (y)))
+#endif
+
+/*
+ * __swap_bytes --
+ *     Swap the contents of two arbitrary values of a given size.
+ */
+static inline void
+__swap_bytes(uint8_t *a, uint8_t *b, size_t size)
+{
+    uint8_t tmp;
+
+    while (size > 0) {
+        tmp = *a;
+        *a++ = *b;
+        *b++ = tmp;
+        --size;
+    }
+}
+
+/*
+ * __med3 --
+ *     Find the median of three values, using the user-supplied comparator.
+ */
+static inline void *
+__med3(void *a, void *b, void *c, wt_cmp_t cmp, void *context)
+{
+#ifndef WT_QSORT_R_DEF
+    WT_UNUSED(context);
+#endif
+
+    if (CMP(a, b, context) < 0) {
+        if (CMP(b, c, context) < 0) {
+            return (b);
+        } else {
+            return (CMP(a, c, context) < 0 ? c : a);
+        }
+    } else {
+        if (CMP(b, c, context) > 0) {
+            return (b);
+        } else {
+            return (CMP(a, c, context) < 0 ? a : c);
+        }
+    }
+}
+
+/*
+ * __qsort --
+ *     Sort an array using a user-supplied comparator function and a context argument.
+ */
+static void
+__qsort(void *arr, size_t nmemb, size_t elem_sz, wt_cmp_t cmp, void *context)
+{
+    /*
+     * This is originally from "Engineering a Sort Function", by Jon L. Bentley and Douglas
+     * M. McIlroy ("Software - Practice and Experience, Vol. 23(11)", November 1993). It's been
+     * massaged into a form suitable for WiredTiger, initially by Luke Pearson.
+     *
+     * It's not simple, so some description of where all these variables point is in order:
+     * -----------------------------------------------
+     * | = |        <         | ? |      >       | = |
+     * -----------------------------------------------
+     *      ^                  ^ ^              ^
+     *      |                  | |              |
+     *      lowest_lt_median   | hi_unknown     highest_gt_median
+     *                         lo_unknown
+     */
+    bool swapped;
+    int cmp_result;
+    size_t lhs_unsorted, rhs_unsorted;
+    uint8_t *a, *hi_pseudomedian, *hi_unknown, *highest_gt_median, *lo_pseudomedian, *lo_unknown,
+      *lowest_lt_median, *pseudomedian;
+
+    WT_ASSERT(NULL, cmp != NULL);
+    if (UNLIKELY(nmemb < 2))
+        return;
+
+    a = arr; /* Avoid arithmetic on void pointers. */
+
+    for (;;) {
+        swapped = false;
+
+        /* Insertion sort for small arrays. */
+        if (nmemb < WT_QSORT_INSERT_THRESHOLD) {
```

Comment:
Interestingly we do define an insertion sort in `misc.h`. Do you think we should try and integrate it? It does feel risky / painful.

Reply:
I had a look, this is indeed painful - it wants the value type passed in (to declare a temporary) and will not play nicely with the "everything is `void *` with a byte count" approach used here.